ksqldb-udf/src/main/java/io/confluent/ksql/schema/ksql/types/SqlLambda.java

```diff
@@ -41,12 +42,27 @@ public static SqlLambda of(
     return new SqlLambda(inputType, returnType);
   }
 
+  public static SqlLambda of(
+      final Integer numInputs
+  ) {
+    return new SqlLambda(numInputs);
+  }
+
   public SqlLambda(
       final List<SqlType> inputTypes,
       final SqlType returnType
   ) {
     this.inputTypes = ImmutableList.copyOf(requireNonNull(inputTypes, "inputType"));
     this.returnType = requireNonNull(returnType, "returnType");
+    this.numInputs = inputTypes.size();
+  }
+
+  public SqlLambda(
+      final Integer numInputs
+  ) {
+    this.inputTypes = null;
+    this.returnType = null;
+    this.numInputs = numInputs;
   }
 
   public List<SqlType> getInputType() {
```

Comment:
I feel like it's unlikely that we'd ever hit this when we have `null` for `inputType` but do we want to consider doing optionals like we did for `SqlArgument`?

Reply:
I think we should take it a step further, this class is coupling two different things: resolved lambdas and unresolved lambdas. Do we even need the former? If so, let's break it out into two separate classes - if code needs one it's unlikely it needs the other (and it can get the size from using `inputTypes.size()`)

Reply:
I do think we need the first to properly resolve generics when we get the return type, but you're right that we don't need both

Reply:
I broke it out into SqlLambda (which only has information on the num input types) and a child class SqlLambdaResolved (which actually has the input types and return type values)